python/tvm/auto_scheduler/feature.py

```diff
@@ -0,0 +1,241 @@
+    """Unpack the flatten feature (in byte array format) from c++
+
+    Parameters
+    ----------
+    byte_arr: bytearray
+        The two-dimensional feature vector in serialized byte array format
+
+    Returns
+    -------
+    features: np.ndarray
+        Feature vectors
+    normalized_throughputs: np.ndarray
+        Normalized throughputs
+    task_ids: np.ndarray
+        Task ids
+    """
+    size_of_int = 4
+    size_of_float = 4
+
+    # The format for n records is:
+    # {
+    #   int n;
+    #   int[n+2] sizes
+
+    #   float[sizes[0]]    feature for record 1
+    #   float[sizes[1]]    feature for record 2
+    #   ...                feature for record i...
+    #   float[sizes[n-1]]  feature for record n
+
+    #   float[sizes[n]]    normalized throughput for n records
+    #   int[sizes[n+1]]    task id for n records
+    # }
+
+    vec_len = DEFAULT_FEATURE_VEC_LEN
+
+    # unpack sizes
+    offset = 0
+    n = struct.unpack_from("1i", byte_arr, offset=offset)[0]
+    offset += size_of_int
+
+    sizes = struct.unpack_from("%di" % (n+2), byte_arr, offset=offset)
+    offset += size_of_int * (n+2)
+
+    # unpack features
+    features = []
+    for size in sizes[:-2]:
+        row = []
+
+        # Now, we need to unpack the feature for multiple statements.
+        # The format is:
+        # {
+        #     int n_stmts
+        #     float[n_stmt][vec_len] feature_vecs
+        # }
+        # where vec_len can be calculated by `(size - 1) / n_stmts`
+
+        if size == 0:
+            # failed during lowering
+            features.append(np.zeros((1, vec_len)))
+        else:
+            n_stmts = struct.unpack_from("f", byte_arr, offset=offset)
+            offset += size_of_float
+
+            n_stmts = int(n_stmts[0] + 0.5)
+            tmp_vec_len = (size - 1) // n_stmts
+            assert tmp_vec_len == vec_len, "The lenght of feature vector is wrong. " \
+                                           "Expected %d but got %d." % (vec_len, tmp_vec_len)
+            assert (size - 1) % n_stmts == 0
+            for _ in range(n_stmts):
+                x = struct.unpack_from("%df" % vec_len, byte_arr, offset=offset)
+                offset += vec_len * size_of_float
+                row.append(x)
+
+            features.append(np.array(row))
+
+    # unpack normalized_throughputs
+    m = sizes[-2]
+    normalized_throughputs = struct.unpack_from("%df" % m, byte_arr, offset=offset)
+    offset += m * size_of_int
+
+    # unpack task_ids
+    m = sizes[-1]
+    task_ids = struct.unpack_from("%di" % m, byte_arr, offset=offset)
+    offset += m * size_of_int
+
+    assert offset == len(byte_arr), "%d vs %d" % (offset, len(byte_arr))
+    return np.array(features, dtype=object), np.array(normalized_throughputs), np.array(task_ids)
+
+
+def get_per_store_features_from_file(filename: str,
+                                     n_lines: int,
+                                     max_n_bufs: int = None) \
```

Comment:
The type of `max_n_bufs` should be `Optional[int]`. Ditto to the rest functions in this file.